### PR TITLE
Cleanup of all the React reconciler logic + remove mutations on ReactElements

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

This PR is a general spring cleaning of all the React logic. Nothing is fixed, but the logic for many of the aspects of the React reconciliation have been simplified and tidied up. 

- I've removed all cases of mutating ReactElements, we now create new ReactElements instead of mutating – because, well, ReactElements are immutable (as are props).
- Removed ReactSerializerState and moved the logic that previous existed into `realm.react`.
- Simplified the branching logic of the reconciler so we no longer need to pass around state and instead use the existing conditional logic.

Manually tested on all our internal bundles are we have no failures.

### CODE REVIEW

File: src/react/branching.js [hunk 1]
```diff
@@ -39,78 +39,74 @@ export type BranchStatusEnum = "ROOT" | "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
...
+          if (xChildren instanceof Value && yChildren instanceof Value) {
+            return findFirstMatchingComponentTypes(xChildren, yChildren);
+          }
+        }
+      } else if (xType === yType) {
+        return applyBranchedLogicValue(realm, value);
       }
-      return true;
+    } else if (x instanceof ArrayValue && y instanceof ArrayValue) {
+      // TODO maybe we nned to support this eventually?
```
Comment (gaearon) [inline src/react/branching.js]: Needs an invariant? It's not clear what happens in that case now.

File: src/react/reconcilation.js [hunk 2]
```diff
@@ -1070,16 +1000,19 @@ export class Reconciler {
 
           for (let [key, binding] of propsValue.properties) {
             if (binding && binding.descriptor && binding.descriptor.enumerable && key !== "children") {
-              setProperty(newProps, key, getProperty(this.realm, propsValue, key));
+              Properties.Set(this.realm, newProps, key, getProperty(this.realm, propsValue, key), true);
```
Comment (gaearon) [inline src/react/reconcilation.js]: Why this change? I thought we intentionally used "internal" methods in these cases

File: src/react/branching.js [hunk 3]
```diff
@@ -39,78 +39,74 @@ export type BranchStatusEnum = "ROOT" | "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
...
-  captureBranchedValue(type: Value, value: Value): Value {
-    this._branchesToValidate.push({ type, value });
-    return value;
+    return realm.evaluateWithAbstractConditional(
+      condValue,
+      () => {
+        return realm.evaluateForEffects(
+          () => applyBranchedLogicValue(realm, consequentVal),
+          null,
+          "splitReactElementsByConditionalType consequent"
```
Comment (gaearon) [inline src/react/branching.js]: Copypaste?

File: src/react/elements.js [hunk 4]
```diff
@@ -109,6 +110,11 @@ function createPropsObject(
     // do we really need to create an object just for { children }
     if (children !== undefined) {
       let childrenObject = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+      // if we have default props, then we need to ensure the children
+      // have keys if they are an array as we're spreading them
+      if (defaultProps !== realm.intrinsics.undefined) {
+        children = applyBranchedLogicValue(realm, children);
```
Comment (gaearon) [inline src/react/elements.js]: Is this fixing a bug? Which one?

I don't see this in the old code

File: src/react/branching.js [hunk 5]
```diff
@@ -39,78 +39,74 @@ export type BranchStatusEnum = "ROOT" | "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
...
 // longer new
-export class BranchState {
-  constructor() {
-    this._branchesToValidate = [];
-  }
-  _branchesToValidate: Array<{
-    type: Value,
-    value: Value,
-  }>;
+export function getValueWithBranchingLogicApplied(realm: Realm, value: AbstractValue): Value {
```
Comment (gaearon) [inline src/react/branching.js]: Can you explain this in more detail on an example? I never really understood what "applying branching logic" means in practice

Comment (trueadm) [reply src/react/branching.js]: It means to add keys to ReactElement nodes in this specific branch.

File: src/react/elements.js (hunk 4 above)
Comment (trueadm) [reply src/react/elements.js]: It was something missed off, not sure of any bugs, but this is the right logic to do after speaking with @sebmarkbage.

File: src/react/reconcilation.js (hunk 2 above)
Comment (trueadm) [reply src/react/reconcilation.js]: After speaking to @NTillmann it seems we don't need to when dealing with ObjectValue setting. So I'd rather remove duplicate logic where possible.

File: src/react/branching.js (hunk 1 above)
Comment (trueadm) [reply src/react/branching.js]: This occurs frequently, but I've not seen anything fail in snapshots. Maybe you can think of a test case?

File: src/react/branching.js (hunk 5 above)
Comment (trueadm) [reply src/react/branching.js]: I also added a comment in my commit.

File: src/react/elements.js (hunk 4 above)
Comment (trueadm) [reply src/react/elements.js]: I tried to come up with a repro, but couldn't make one that fails. If you can think of one, let me know. From memory, I think it's to do with key checks on children from props vs explicit children.

Comment (gaearon) [reply src/react/elements.js]: This needs a test case then?

**Comment by gaearon** [comment]
It's hard for me to tell what is just cleanup and what is fixing bugs. Can you separate the two, and make test cases for changes that affect the output?

File: src/react/branching.js [hunk 6]
```diff
@@ -39,78 +39,74 @@ export type BranchStatusEnum = "ROOT" | "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
...
-      }
-    }
-    if (applyBranchedLogic) {
-      for (let i = 0; i < this._branchesToValidate.length; i++) {
-        this._applyBranchedLogicValue(realm, reactSerializerState, this._branchesToValidate[i].value);
+          if (xChildren instanceof Value && yChildren instanceof Value) {
+            return findFirstMatchingComponentTypes(xChildren, yChildren);
+          }
+        }
+      } else if (xType === yType) {
```
Comment (gaearon) [inline src/react/branching.js]: Per our discussion let's fix the bug here and add a regression test.
My sketch:

```js
var React = require('react');
// the JSX transform converts to React, so we need to add it back in
this['React'] = React;

function App(props) {
  if (props.foo) {
    return <Foo />;
  }
  return <Bar />;
}

function Foo() {
  return <input />;
}

function Bar() {
  return <input />;
}

if (this.__optimizeReactComponentTree) {
  __optimizeReactComponentTree(App);
}

module.exports = App;
```

**Comment by trueadm** [comment]
@gaearon I added a ton of test cases and handled all the different possible cases.

File: test/react/functional-components/type-change2.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  }
+  renderer.update(<Root callback={callback} foo={true} />);
+  renderer.update(<Root callback={callback} foo={false} />);
+
+  let results = [];
+  results.push(['ensure refs was called 3 times', counter === 3]);
+  results.push(['ensure refs at 0 is not null', nodes[0] !== null]);
+  results.push(['ensure refs at 1 is null', nodes[1] === null]);
+  results.push(['ensure refs at 2 is not null', nodes[2] !== null]);
+  return results;
```
Comment (gaearon) [inline test/react/functional-components/type-change2.js]: We need to also ensure `nodes[0] !== nodes[2]`.

File: src/react/branching.js [hunk 8]
```diff
@@ -80,10 +80,31 @@ export function getValueWithBranchingLogicApplied(
...
+        let xElem;
+        if (x instanceof ArrayValue) {
+          // handle the case of [y].equals([x]
+          xElem = getProperty(realm, y, index + "");
+        } else if (index === 0) {
+          // handle the case of [y].equals(x)
+          xElem = x;
+        }
 
         if (xElem instanceof Value && yElem instanceof Value) {
```
Comment (gaearon) [inline src/react/branching.js]: Note it’ll also be important that only one level nesting is special cased. Deeper nesting change loses state.

File: src/react/branching.js [hunk 9]
```diff
@@ -39,78 +39,132 @@ export type BranchStatusEnum = "ROOT" | "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
...
+      } else if (!xType.equals(yType)) {
+        return findMatchingComponentTypes(xType, yType);
       }
-      return true;
+    } else if (
+      ArrayValue.isIntrinsicAndHasWidenedNumericProperty(x) ||
+      ArrayValue.isIntrinsicAndHasWidenedNumericProperty(y)
+    ) {
+      // If either case is an unknown array, we do not know
+      // the contents of the array, so we cannot add keys
```
Comment (gaearon) [inline src/react/branching.js]: And is it safe to skip then? Why? Adding keys seems essential for correctness.

File: src/react/branching.js [hunk 10]
```diff
@@ -39,78 +39,132 @@ export type BranchStatusEnum = "ROOT" | "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
 // objects depending on various heuristics (if they have the same "type" for example)
 // A new branch state is created on a branch status of "NEW_BRANCH" and is reset to null once the branch is no
```
Comment (gaearon) [inline src/react/branching.js]: ^^ this comment seems completely irrelevant now. Could you please make sure that when the code changes, the comments are still helpful?

File: src/react/branching.js (hunk 9 above)
Comment (trueadm) [reply src/react/branching.js]: Out of scope for now. We definitely didn't support this before and this PR was more of a clean up than adding brand new functionality. We might be able to add some form of helper method at runtime that adds in keys – as the array is completely abstract to us.

Comment (gaearon) [reply src/react/branching.js]: Yeah, but it seems like in that case we should fail hard instead of emitting code that works differently.

Comment (trueadm) [reply src/react/branching.js]: Let's handle in a follow up still. Failing hard here may make sense once we have more tests around it. I've updated https://github.com/facebook/prepack/issues/1131 to include a point about them.

File: src/invariant.js [hunk 11]
```diff
@@ -11,6 +11,7 @@
 
 export default function invariant(condition: boolean, format: string): void {
   if (condition) return;
+  return;
```
Comment (gaearon) [inline src/invariant.js]: Oops
